var/spack/repos/builtin/packages/adios2/package.py

```diff
@@ -281,3 +283,50 @@ def setup_run_environment(self, env):
             env.prepend_path("HDF5_PLUGIN_PATH", os.path.dirname(all_libs[idx]))
         except ValueError:
             pass
+
+    @run_after("install")
+    def setup_install_tests(self):
+        """
+        Copy the example files after the package is installed to an
+        install test subdirectory for use during `spack test run`.
+        """
+        extra_install_tests = [join_path("testing", "install", "C")]
+        self.cache_extra_test_sources(extra_install_tests)
+
+    def test_run_executables(self):
+        """Run installed adios2 executables"""
+
+        commands_and_args = [("bpls", ["-v", "-V"])]
+
+        for cmd, opts in commands_and_args:
+            with test_part(
+                self, f"test_adios2_cmds_{cmd}", purpose=f"run installed adios2 executable {cmd}"
+            ):
+                exe = which(join_path(self.prefix.bin, cmd))
+                exe(*opts)
+
+    def test_build_example(self):
+        """Build and run an example program"""
+        src_dir = self.test_suite.current_test_cache_dir.testing.install.C
+        test_stage_dir = self.test_suite.test_dir_for_spec(self.spec)
+
+        # Create the build tree within this spec's test stage dir so it gets
+        # cleaned up automatically
+        build_dir = tempfile.mkdtemp(dir=test_stage_dir)
+
+        std_cmake_args = []
+
+        if "+mpi" in self.spec:
+            mpi_exec = join_path(self.spec["mpi"].prefix, "bin", "mpiexec")
+            std_cmake_args.append("-DMPIEXEC_EXECUTABLE={0}".format(mpi_exec))
+
+        built_programs = ["adios_c_mpi_test", "adios_adios2c_test", "adios_c_test"]
+
+        with working_dir(build_dir):
+            cmake(src_dir, *std_cmake_args)
+            make()
+
+            for example in [join_path(build_dir, p) for p in built_programs]:
+                exe = which(example)
+                if exe:
+                    exe()
```

Comment:
Curious.  Why not use `test_part` here (like you did starting line 302)? so the results of running each built program will be reported separately?

Reply:
Makes sense, I'll fix this.